src/main/scala/uk/gov/hmrc/ct/box/SelfValidatableBox.scala

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2016 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.ct.box
+
+import org.joda.time.LocalDate
+import uk.gov.hmrc.ct.box.retriever.BoxRetriever
+import uk.gov.hmrc.ct.ct600.v3.retriever.RepaymentsBoxRetriever
+import uk.gov.hmrc.ct.domain.ValidationConstants._
+import uk.gov.hmrc.ct.utils.DateImplicits._
+import ValidatableBox._
+
+
+trait SelfValidatableBox[T <: BoxRetriever, B] extends Validators with ValidatableBox[T]{
+
+  box: CtValue[B] with CtBoxIdentifier =>
+
+  def validate(boxRetriever: T): Set[CtValidation]
+
+  protected def validateBooleanAsMandatory()()(implicit ev: <:<[B, Option[Boolean]]): Set[CtValidation] = {
```

Comment:
the implicit evidence makes the methods type safe, catching errors at compile time and preventing class cast exceptions at runtime